datamade/just-spaces
# Grouped Observational Age plugin

## Overview

This PR implements a custom PLDP age widget for observational studies as described in #105. It uses Django's [MultiValueField](https://docs.djangoproject.com/en/2.2/ref/forms/fields/#multivaluefield) and [MultiWidget](https://docs.djangoproject.com/en/2.2/ref/forms/widgets/#multiwidget), which work pretty smoothly EXCEPT I spent a while spinning my wheels on how to add labels to each subwidget—ie, the age ranges as seen here:

<img width="761" alt="Screen Shot 2019-04-29 at 2 26 41 PM" src="https://user-images.githubusercontent.com/1094243/56921233-dbf7ec00-6a8a-11e9-98c9-ac238fd36079.png">

I think these sub-labels are necessary, but I'm not pleased with my method right now as it involves editing rendered HTML. (See [here](https://github.com/datamade/just-spaces/blob/issue-105/fobi_custom/plugins/form_elements/fields/observational/widgets.py#L16-L25)). It works, but I wonder if it could be better. This is something that [`format_output`](https://cdf.9vo.lt/1.10/django.forms.widgets/MultiWidget.html#format_output) could have handled before it was deprecated.

This widget currently saves data as a list of tuples. For example, the image below would turn into the following list.

<img width="732" alt="Screen Shot 2019-04-29 at 2 38 06 PM" src="https://user-images.githubusercontent.com/1094243/56921855-7442a080-6a8c-11e9-920d-5b3057c7191a.png">

`[('0-14', 4), ('15-24', 10), ('25-64', 100), ('65+', 17)]`

This is a starting point that will be straightforward to change. @jeancochrane, let's chat about how this data should be formatted in order to connect smoothly with your visualizations.

Once we've merged this PR I'll still need to create the other observational elements described by the PLDP. However, I've made `ObservationalField` and `ObservationalWidget` reusable so this PR should include the heaviest technical lift for #105.

## Testing Instructions

 * If you don't already have it, create the database: `pg_restore -C -j4 --no-owner just-spaces.dump | psql`
* Migrate: `python manage.py migrate`
* Login as testuser (password in the DataMade LastPass under "Just Spaces Staging")
* Create a new Observational survey
* Add an Age element to it. Can you change the detail level, and does the resulting element look like you'd expect?

Connects #105
